source/common/upstream/health_checker_impl.h

```diff
@@ -417,5 +421,87 @@ class RedisHealthCheckerImpl : public HealthCheckerImplBase {
   Redis::ConnPool::ClientFactory& client_factory_;
 };
 
+/**
+ * gRPC health checker implementation.
+ */
+class GrpcHealthCheckerImpl : public HealthCheckerImplBase {
+public:
+  GrpcHealthCheckerImpl(const Cluster& cluster, const envoy::api::v2::HealthCheck& config,
+                        Event::Dispatcher& dispatcher, Runtime::Loader& runtime,
+                        Runtime::RandomGenerator& random);
+
+private:
+  struct GrpcActiveHealthCheckSession : public ActiveHealthCheckSession,
+                                        public Http::StreamDecoder,
+                                        public Http::StreamCallbacks {
+    GrpcActiveHealthCheckSession(GrpcHealthCheckerImpl& parent, HostSharedPtr host);
+    ~GrpcActiveHealthCheckSession();
+
+    void onRpcComplete(Grpc::Status::GrpcStatus grpc_status, const std::string& grpc_message,
+                       bool end_stream);
+    bool isHealthCheckSucceeded(Grpc::Status::GrpcStatus grpc_status) const;
+    void resetState();
+    void logHealthCheckStatus(Grpc::Status::GrpcStatus grpc_status,
+                              const std::string& grpc_message);
+
+    // ActiveHealthCheckSession
+    void onInterval() override;
+    void onTimeout() override;
+
+    // Http::StreamDecoder
+    void decodeHeaders(Http::HeaderMapPtr&& headers, bool end_stream) override;
+    void decodeData(Buffer::Instance&, bool end_stream) override;
+    void decodeTrailers(Http::HeaderMapPtr&&) override;
+
+    // Http::StreamCallbacks
+    void onResetStream(Http::StreamResetReason reason) override;
+    void onAboveWriteBufferHighWatermark() override {}
+    void onBelowWriteBufferLowWatermark() override {}
+
+    void onEvent(Network::ConnectionEvent event);
+
+    class ConnectionCallbackImpl : public Network::ConnectionCallbacks {
+    public:
+      ConnectionCallbackImpl(GrpcActiveHealthCheckSession& parent) : parent_(parent) {}
+      // Network::ConnectionCallbacks
+      void onEvent(Network::ConnectionEvent event) override { parent_.onEvent(event); }
+      void onAboveWriteBufferHighWatermark() override {}
+      void onBelowWriteBufferLowWatermark() override {}
+
+    private:
+      GrpcActiveHealthCheckSession& parent_;
+    };
+
+    ConnectionCallbackImpl connection_callback_impl_{*this};
+    GrpcHealthCheckerImpl& parent_;
+    Http::CodecClientPtr client_;
```

Comment:
Before going into the review, I'd like to understand some design stuff. How come we're doing all the low level gRPC codec and connection management in this HC rather than leveraging the existing gRPC client support in Envoy (https://github.com/envoyproxy/envoy/blob/master/include/envoy/grpc/async_client.h, https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/async_client_impl.h)?

Reply:
grpc AsyncClientImpl was my first thought. Unfortunately async client is tied to cluster manager, router, retries, AsyncHttpClient, etc. We don't have (and don't need) all that abstractions in healthchecker code - we need to connect directly to a specific endpoint.

Other option is to refactor out connection establishment part from current Grpc::AsyncStreamImpl, leaving it as a dumb protocol decoder, but that requires careful thinking and greatly widens scope of this pretty germetic PR, which I would like to avoid and work on tech debt later. Imo, low-level protocol stuff here is pretty simple.

Reply:
Got it, I think this is fine then, will take a deeper look in a bit.

Reply:
FWIW the issue of doing it this way vs. AsyncClient came up when @craffert0 took a quick stab at implementing this a couple of months ago. I think it's definitely possible to refactor AsyncClient so it doesn't depend on cluster manager, but it's non-trivial. I'm in favor of doing it this way and adding a TODO to come back and revisit at a later time.